python/pyarrow/tests/test_substrait.py

```diff
@@ -0,0 +1,104 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+import os
+import sys
+import pytest
+
+import pyarrow as pa
+from pyarrow.lib import tobytes
+from pyarrow.lib import ArrowInvalid
+
+try:
+    import pyarrow.parquet as pq
+except ImportError:
+    pq = None
+
+try:
+    import pyarrow.substrait as substrait
+except ImportError:
+    substrait = None
+
+# Marks all of the tests in this module
+# Ignore these with pytest ... -m 'not substrait'
+pytestmark = [pytest.mark.dataset, pytest.mark.substrait]
+
+
+@pytest.mark.skipif(sys.platform == 'win32',
+                    reason="ARROW-16392: file based URI is" +
+                    " not fully supported for Windows")
+def test_run_serialized_query(tmpdir):
+    substrait_query = """
+    {
+        "relations": [
+        {"rel": {
+            "read": {
+            "base_schema": {
+                "struct": {
+                "types": [
+                            {"i64": {}}
+                        ]
+                },
+                "names": [
+                        "foo"
+                        ]
+            },
+            "local_files": {
+                "items": [
+                {
+                    "uri_file": "file://FILENAME_PLACEHOLDER"
+                }
+                ]
+            }
+            }
+        }}
+        ]
+    }
+    """
+
+    schema = pa.schema([pa.field('foo', pa.int64())])
+    path = os.path.join(str(tmpdir), 'substrait_data.arrow')
+    with pa.OSFile(path, 'wb') as sink:
+        with pa.ipc.new_file(sink, schema) as writer:
+            batch = pa.record_batch(
+                [pa.array(range(5), type=pa.int64())], schema)
+            writer.write(batch)
+
+    with pa.OSFile(path, 'rb') as source:
+        expected_table = pa.ipc.open_file(source).read_all()
+
+    query = tobytes(substrait_query.replace("FILENAME_PLACEHOLDER", path))
+
+    buf = pa._substrait._parse_json_plan(query)
+
+    reader = substrait.run_query(buf)
+    res_tb = reader.read_all()
+
+    assert expected_table.num_rows == res_tb.num_rows
```

Comment:
Can you check the table contents instead of only the number of rows?

Reply:
Ah yes, I think we can compare the content in the data array. 

Looking into the two tables (first the output from substrait, second the table we created)

Table from Substrait

```bash
pyarrow/tests/test_substrait.py pyarrow.Table
foo: int64
__fragment_index: int32
__batch_index: int32
__last_in_fragment: bool
__filename: string
----
foo: [[1,2,3,4,5]]
__fragment_index: [[0,0,0,0,0]]
__batch_index: [[0,0,0,0,0]]
__last_in_fragment: [[true,true,true,true,true]]
__filename: [["/private/var/folders/vq/2v5vdv2n1t9b_fp4lp17h3d40000gn/T/pytest-of-vibhatha/pytest-48/test_run_serialized_query0/substrait_data.arrow","/private/var/folders/vq/2v5vdv2n1t9b_fp4lp17h3d40000gn/T/pytest-of-vibhatha/pytest-48/test_run_serialized_query0/substrait_data.arrow","/private/var/folders/vq/2v5vdv2n1t9b_fp4lp17h3d40000gn/T/pytest-of-vibhatha/pytest-48/test_run_serialized_query0/substrait_data.arrow","/private/var/folders/vq/2v5vdv2n1t9b_fp4lp17h3d40000gn/T/pytest-of-vibhatha/pytest-48/test_run_serialized_query0/substrait_data.arrow","/private/var/folders/vq/2v5vdv2n1t9b_fp4lp17h3d40000gn/T/pytest-of-vibhatha/pytest-48/test_run_serialized_query0/substrait_data.arrow"]]
```

Table 
```bash
pyarrow.Table
foo: int64
----
foo: [[1,2,3,4,5]]
```

Reply:
But we cannot compare the two table objects, right? The tables are created in different ways, one in-memory and one by reading from the disk, so the metadata will be different right? 
cc @westonpace

Reply:
The augmented columns will be a problem but otherwise we should be able to compare them.  The test is not adding any metadata and the exec plan strips out all metadata so I don't expect a metadata conflict.  So just drop the extra columns using [`pa.Table.select`](https://arrow.apache.org/docs/python/generated/pyarrow.Table.html#pyarrow.Table.select).

Reply:
The R implementation ran into this same issue.  We should not be adding the extra columns since the Substrait plan is not asking for them.

Reply:
Also, to be clear, I'm not suggesting you add anything to the implementation itself, only the tests.  There is already a follow-up PR to address the implementation.

Reply:
I understand. Clear!